core/chains/evm/logpoller/log_poller.go

```diff
@@ -1121,13 +1142,39 @@ func (lp *logPoller) PruneOldBlocks(ctx context.Context) (bool, error) {
 	return lp.logPrunePageSize == 0 || rowsRemoved < lp.logPrunePageSize, err
 }
 
-// PruneExpiredLogs logs that are older than their retention period defined in Filter.
-// Returns whether all logs eligible for pruning were removed. If logPrunePageSize is set to 0, it will always return true.
+// PruneExpiredLogs will attempt to remove any logs which have passed their retention period. Returns whether all expired
+// logs were removed. If logPrunePageSize is set to 0, it will always return true unless an actual error is encountered
 func (lp *logPoller) PruneExpiredLogs(ctx context.Context) (bool, error) {
+	done := true
+
 	rowsRemoved, err := lp.orm.DeleteExpiredLogs(ctx, lp.logPrunePageSize)
-	return lp.logPrunePageSize == 0 || rowsRemoved < lp.logPrunePageSize, err
+	if err != nil {
+		lp.lggr.Errorw("Unable to find excess logs for pruning", "err", err)
+		return false, err
+	} else if lp.logPrunePageSize != 0 && rowsRemoved == lp.logPrunePageSize {
+		done = false
+	}
+
+	if !lp.countBasedLogPruningActive.Load() {
```

Comment:
Why do we need this check? Are we concerned about `SelectExcessLogIDs` query performance when nothing is matched? Ideally when there are no filters with `MaxLogsKept>0` the `SelectExcessLogIDs` should be insanely fast.

If we have a flag for `countBasedLogPruningActive` why don't we have a flag for `timeRetentionBasedLogPruningActive`? ;)

Reply:
> Are we concerned about SelectExcessLogIDs query performance when nothing is matched? Ideally when there are no filters with MaxLogsKept>0 the SelectExcessLogIDs should be insanely fast.

I don't know of any easy way of making the query any faster for the case where they're all zero, aside from just adding an extra query before it to check for that special case.  We can't just ignore rows where MaxLogKept=0 because each of them affects all other rows (if any log joins with both that one and another row with MaxLogsKept > 0 then we can't drop it). Only when they happen to be all 0's can any of them be ignored... which is the kind of thing the planner is just not designed to figure out.

So one way to think about the purpose of this `lp.countBasedLogPruneActive` flag is, it's just an optimization so that we don't have to add an extra query to check for that special case every time.

The same is true for the retention query, but that one is by nature a lot of faster because it doesn't have to sort and count subsets of logs... it only has to do a single comparison.  So it's less savings, plus it's already in use... so I'm not sure it would matter any more--the most cpu/memory savings would be now when the feature is first being introduced and nobody is using it yet. It also gives us something fairly similar to a "feature flag" where we know that it can be rolled out without hurting anything while it continues to get tested in new ways as products start to make use of it on develop branches.

I wanted to confirm all of the above by actually load testing it, but for the past week the load tests have been broken on `ccip-develop` branch (until today, I kept thinking it must have been my change, but apparently it was due to ChainReader getting enabled in the docker file). Tonight I finally got the load test to work again, by rebasing onto an earlier commit in ccip repo from a couple weeks ago. The test confirms that the query times for SelectExcessLogIDs are about the same (maybe slightly faster, but not a significant difference) with all MaxLogKept=0 rows as when some rows are non-zero.

So I've left it in there, and addressed @dhaidashenko 's comment about having not fully implemented it the previous time.